src/Microsoft.AspNet.Mvc.Xml/RequiredValidationHelper.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Reflection;
+using System.Runtime.Serialization;
+using Microsoft.AspNet.Mvc.ModelBinding;
+
+namespace Microsoft.AspNet.Mvc.Xml
+{
+    /// <summary>
+    /// Validates types having value type properties decorated with <see cref="RequiredAttribute"/>
+    /// but no <see cref="DataMemberAttribute"/>.
+    /// </summary>
+    public static class RequiredValidationHelper
+    {
+        private static ConcurrentDictionary<Type, Dictionary<Type, List<string>>> cachedValidationErrors
+            = new ConcurrentDictionary<Type, Dictionary<Type, List<string>>>();
+
+        public static void Validate([NotNull] Type modelType, [NotNull] ModelStateDictionary modelStateDictionary)
+        {
+            var visitedTypes = new HashSet<Type>();
+
+            // Every node maintains a dictionary of Type => Errors. 
+            // It's a dictionary as we want to avoid adding duplicate error messages.
+            // Example:
+            // Type 'A' has a child node of type 'B' somewhere in the left sub-tree
+            // and it has type 'B' again  somewhere down the righ sub-tree. From the perspective of
+            // type 'A', it should not have duplicate error messgaes for type 'B'.
+            var rootNodeValidationErrors = new Dictionary<Type, List<string>>();
```

Comment:
it would be simpler just to type out an example. there's no such thing as 'right subtree' as this is an n-ary tree